File: legacy/DVR3d.m
%This code computes the energy levels for a single tweezer.

%Fundamental constants
% constants;
h = 6.62607004 * 1e-34;
hbar = h / (2 * pi);
kb = 1.38064852 * 1e-23;

% Lithium-6
mLi = 6.015122 * 1.6605339 * 1e-27;
gamma = 2 * pi * 6 * 1e6;
lambda0 = 671 * 1e-9;
c = 299792458;
f0 = c / lambda0;
omega0 = 2 * pi * f0;

% Laser Params
lambda = 780 * 1e-9; %wavelength (m)
omega = 2 * pi * c / lambda;
k = 2 * pi / lambda;
Er = hbar^2 * k^2 / (2 * mLi);

P = 55 * 1e-6; %power per tweezer (W)
asymm = 0; %asymmetry in units of depth
wx = 1000 * 1e-9; %waist perp direction (m) 1350
wy = 1000 * 1e-9; %waist tunneling direction (m) 1000% 810e-9
zRx = pi * wx^2 / lambda;
zRy = pi * wy^2 / lambda;
%
I0 = 2 * P / (pi * wx * wy);
V0 = 3 * pi * c^2 / (2 * omega0^3) * (gamma / (omega0 - omega) + gamma / (omega0 + omega)) * I0;

V_tweezers = @(x, y, z) -V0 ./ (1 + z.^2 .* (1 ./ zRx.^2 + 1 ./ zRy.^2) / 2) ...
    .* exp(-2 * x.^2 ./ (wx.^2 .* (1 + z.^2 ./ zRx.^2))) ...
    .* exp(-2 * y.^2 ./ (wy.^2 .* (1 + z.^2 ./ zRy.^2)));

% Gravity
g = 9.80665;
V_gravity = @(x, y, z) mLi * g * z;

% V = @(x, y, z) V_tweezers(x, y, z) + V_gravity(x, y, z);
V = @(x, y, z) V_tweezers(x, y, z);

% Tweezer Params
M = [2 * wx 2 * wy 4 * max(zRx, zRy)];
% N = [15 15 31];
Nmax = 8;
E = [];

for index = 1:Nmax
    Ni = 2 * (index * [1, 1, 2] + [0 0 1]) + 1;
    E(end + 1, :) = DVR3d_run(M, Ni, hbar, mLi, V)'/V0;
end

dE = abs(diff(E));
close all

for index = 1:5
    plot((2:Nmax)', dE(:, index))
    hold on
end

xlabel("N")
ylabel("\Delta E/V_0")
set(gca, "YScale", 'log')
title(['Convergence of energies vs N @ Tweezer, fixing R = ', num2str(M(1) / 2 / wx), 'w'])

function E = DVR3d_run(M, N, hbar, mLi, V)
    %DVR3d - Description
    %
    % Syntax: E = DVR3d_run(M, N, hbar, mLi, V)
    %
    % Long description

    tic;
    % Tweezer Params
    M = num2cell(M);
    N = num2cell(N);
    [Mx, My, Mz] = deal(M{:});
    [Nx, Ny, Nz] = deal(N{:});

    % Discretization params
    x = linspace(-Mx / 2, Mx / 2, Nx);
    y = linspace(-My / 2, My / 2, Ny);
    z = linspace(-Mz / 2, Mz / 2, Nz);
    dx = x(2) - x(1);
    dy = y(2) - y(1);
    dz = z(2) - z(1);
    [xx, yy, zz] = meshgrid(x, y, z);
    fprintf('Initialized parameters. Elapsed Time %.3f s\n', toc)

    %get potential
    VV = V(xx, yy, zz);
    VV = spdiags(VV(:), 0, Nx * Ny * Nz, Nx * Ny * Nz);
    fprintf('Set up potential. Elapsed Time %.3f s\n', toc)

    %get kinetic DVR term
    TTx = spdiags(ones(Nx * Ny * Nz, 1) * pi^2/3, 0, Nx * Ny * Nz, Nx * Ny * Nz); % diagonal term

    % build off-diagonals in matrix form
    for Dx = 1:(Nx - 1)
        M = [ones(1, (Nx - Dx) * Ny), zeros(1, Ny * Dx)]; % [1 1 1 ... 1 0 0 ...]
        M = repmat(M, 1, Nz)'; % repeat by Nz times
        TTx = spdiags((-1)^Dx * 2 / Dx^2 * M(end:-1:1), +Dx * Ny, TTx); % replace Dx * Ny-diagonals of TTx by (-1)^Dx * 2 / Dx^2 * M(end:-1:1)
        TTx = spdiags((-1)^Dx * 2 / Dx^2 * M, -Dx * Ny, TTx); % the symmetric entries
    end

    TTy = spdiags(ones(numel(yy), 1) * pi^2/3, 0, Nx * Ny * Nz, Nx * Ny * Nz);

    for Dy = 1:(Ny - 1)
        M = [ones(1, Ny - Dy), zeros(1, Dy)];
        M = repmat(M, 1, Nx * Nz)';
        TTy = spdiags((-1)^Dy * 2 / Dy^2 * M(end:-1:1), +Dy, TTy);
        TTy = spdiags((-1)^Dy * 2 / Dy^2 * M, -Dy, TTy);
    end

    TTz = spdiags(ones(numel(zz), 1) * pi^2/3, 0, Nx * Ny * Nz, Nx * Ny * Nz);

    for Dz = 1:(Nz - 1)
        M = ones(1, Nx * Ny * Nz)';
        TTz = spdiags((-1)^Dz * 2 / Dz^2 * M, +Dz * Nx * Ny, TTz);
        TTz = spdiags((-1)^Dz * 2 / Dz^2 * M, -Dz * Nx * Ny, TTz);
    end

    TT = hbar^2 / (2 * mLi) * (TTx / dx^2 + TTy / dy^2 + TTz / dz^2);
    % TT = (TTx + TTy + TTz);
    fprintf('Kinetic Energy Term. Elapsed Time %.3f s\n', toc)

    %diagonalize
    NN = 16;
    H = TT + VV - spdiags(ones(numel(xx), 1) * min(VV(:)), 0, Nx * Ny * Nz, Nx * Ny * Nz); % subtract zero poin energy
    [eigvects, eigenergies] = eigs(H, NN, 'SM');
    fprintf('Diagonalize. Elapsed Time %.3f s\n', toc)

    % extract energies and sort
    E = diag(eigenergies + min(VV(:)));
    [E, II] = sort(E);
    PSI = eigvects(:, II);

    % rearange psi and normalize
    psi = zeros(Nx, Ny, Nz, NN);

    for ii = 1:NN
        psi_temp = reshape(PSI(:, ii), Ny, Nx, Nz);
        Norm = sum(psi_temp(:).^2) * dx * dy * dz;
        PSI(:, ii) = PSI(:, ii) / sqrt(Norm);
        psi_temp = psi_temp / sqrt(Norm) * sign(max(psi_temp(:)));
        psi(:, :, :, ii) = permute(psi_temp, [2 1 3]);
    end

end

% % close all
% figure(45); clf;
% colors = lines(NN);
% subplot(1, 3, 1)
% hold on
% plot(x * 1e6, V(x, 0, 0) / h * 1e-3, 'k')
% % plot(x*1e6, V(x, a/2, 0)/h*1e-3, 'k')
% for ii = 1:NN
%     plot([min(x), max(x)] * 1e6, [E(ii), E(ii)], '--', 'color', colors(ii, :));
% end

% xlabel('X (\mum)')
% ylabel('Potential (kHz)')
% title('X axis')
% grid on
% xlim([min(x), max(x)] * 1e6)
% ylim([1.1 * min(VV(:)) / h * 1e-3, 0])
% subplot(1, 3, 2)
% hold on
% plot(y * 1e6, V(0, y, 0) / h * 1e-3, 'k')

% for ii = 1:NN
%     plot([min(y), max(y)] * 1e6, [E(ii), E(ii)], '--', 'color', colors(ii, :));
% end

% xlabel('Y (\mum)')
% title('Y axis')
% grid on
% xlim([min(y), max(y)] * 1e6)
% ylim([1.1 * min(VV(:)) / h * 1e-3, 0])
% subplot(1, 3, 3)
% hold on
% plot(z * 1e6, V(0, 0, z) / h * 1e-3, 'k')
% % plot(z*1e6, V(0, a/2, z)/h*1e-3, 'k')
% for ii = 1:NN
%     plot([min(z), max(z)] * 1e6, [E(ii), E(ii)], '--', 'color', colors(ii, :));
% end

% xlabel('Z (\mum)')
% title('Z axis')
% grid on
% xlim([min(z), max(z)] * 1e6)
% ylim([1.1 * min(VV(:)) / h * 1e-3, 0])
% sgtitle(sprintf('Potential and bound state energies\n wx = %.0f nm, wy = %.0f nm, power = %.3f mW', wx * 1e9, wy * 1e9, P * 1e3))

% figure(8); clf;

% for ii = 1:NN
%     subplot(NN, 3, 3 * ii - 2)

%     if ii == 1
%         title('X axis')
%     end

%     ylabel(sprintf('Psi_%d', ii))

%     if ii == NN
%         xlabel('X (\mum)')
%     end

%     hold on

%     for jj = 1:round(Ny / 5):Ny

%         for kk = 1:round(Nz / 5):Nz
%             plot(x * 1e6, squeeze(psi(:, jj, kk, ii)))
%         end

%     end

%     xlim([min(x), max(x)] * 1e6)
%     grid on
%     subplot(NN, 3, 3 * ii - 1)

%     if ii == 1
%         title('Y axis')
%     end

%     if ii == NN
%         xlabel('Y (\mum)')
%     end

%     hold on

%     for jj = 1:round(Nx / 5):Nx

%         for kk = 1:round(Nz / 5):Nz
%             plot(y * 1e6, squeeze(psi(jj, :, kk, ii)))
%         end

%     end

%     xlim([min(y), max(y)] * 1e6)
%     grid on
%     subplot(NN, 3, 3 * ii - 0)

%     if ii == 1
%         title('Z axis')
%     end

%     if ii == NN
%         xlabel('Z (\mum)')
%     end

%     hold on

%     for jj = 1:round(Nx / 5):Nx

%         for kk = 1:round(Ny / 5):Ny
%             plot(z * 1e6, squeeze(psi(jj, kk, :, ii)))
%         end

%     end

%     xlim([min(z), max(z)] * 1e6)
%     grid on
% end

% sgtitle(sprintf('Wavefunctions\n wx = %.0f nm, wy = %.0f nm, power = %.3f mW', wx * 1e9, wy * 1e9, P * 1e3))
% %
% figure(102); clf;
% plot(1:numel(E), E - E(1), 'o');
% ylabel('E (kHz)'); xlabel('Eigenlevels');
% legend(num2str(E - E(1)), 'location', 'best');

File: legacy/DVR3d.py
import numpy as np
import scipy.sparse as sp
import numpy.linalg as la
import scipy.sparse.linalg as sla
from scipy.sparse.linalg import LinearOperator
from opt_einsum import contract
from time import time
import sys

# Fundamental constants
a0 = 5.29E-11  # Bohr radius, in unit of meter
Eha = 6579.68392E12 * 2 * np.pi  # Hartree energy, in unit of Hz
amu = 1822.89  # atomic mass, in unit of electron mass
hb = 1  # Reduced Planck const
m = 6.015122 * amu  # Lithium-6 atom mass, in unit of electron mass
dim = 3  # space dimension

# Experiment parameters in atomic units
V0_SI = 1.0452E5 * 2 * np.pi  # 104.52kHz * h, potential depth, in SI unit, since hbar is set to 1 this should be multiplied by 2pi
V0 = V0_SI / Eha  # potential depth in unit of Hartree energy
# TO GET A REASONABLE ENERGY SCALE, WE SET v=1 AS THE ENERGY UNIT HEREAFTER
w = 1E-6 / a0  # ~1000nm, waist length, in unit of Bohr radius
l = 780E-9 / a0  # 780nm, light wavelength
zR = np.pi * w**2 / l  # ~4000nm, Rayleigh range

Nmax = np.array([7, 7, 15])  # Max number of grid points
dx0 = np.array([1 / 8 * w, 1 / 8 * w, 0.4738 * w])  # Fixed delta x value
# Fixed R value, should be larger than Rayleigh length scale
R0 = np.array([w, w, 2 * zR])
k = 10  # Number of energy levels to track
# TODO:
#       3. Make use of system symmetries, reflection, cylindrical, etc.

## Abosorbers
VI0 = 4E5 / V0_SI  # Absorption potential strength in unit of V0
LI = .3 * w  # Absorption region, in unit of dx

# Harmonic parameters
dx0_sho = 1 / 3 * np.ones(3, dtype=float)
Nmax_sho = 30
R0_sho = Nmax_sho * dx0_sho
omega = 1
m_sho = 1


def Vfun(x, y, z):
    # Potential function, Eq.2 in PRA
    den = 1 + (z / zR)**2
    V = -1 / den * np.exp(-2 * (x**2 + y**2) / (w**2 * den))
    return V


def Vabs(x, y, z, VI=VI0, L=np.inf * np.ones(dim), R=np.zeros(dim)):
    r = np.array([x, y, z]).transpose(1, 2, 3, 0)
    np.set_printoptions(threshold=sys.maxsize)
    d = abs(r) - R
    d = (d > 0) * d
    Vi = np.sum(d / L, axis=3)
    if Vi.any() != 0.0:
        V = -1j * VI * Vi
    return V


def Vsho(x, y, z):
    # Harmonic potential function
    V = m_sho / 2 * omega**2 * (x**2 + y**2 + z**2)
    return V


def Vmat(n, dx, potential, avg=1, absorber=False, ab_param=(LI, VI0)):
    # Potential energy tensor, index order [x y z x' y' z']
    # NOTE: here n, dx are 3-element np.array s.t. n = [nx, ny, nz], dx = [dx, dy, dz]
    #       potential(x, y, z) is a function handle to be processed as potential function for solving
    x = []
    for i in range(dim):
        x.append(np.arange(-n[i], n[i] + 1) * dx[i])
    X = np.meshgrid(x[0], x[1], x[2], indexing='ij')
    # 3 index tensor V(x, y, z)
    V = avg * potential(X[0], X[1], X[2])
    if absorber:  # add absorption layer
        L = ab_param[0] * (n > 0)
        Ri = n * dx - ab_param[0]
        L[np.nonzero(n == 0)] = np.inf
        V = V.astype(complex) + Vabs(
            X[0], X[1], X[2], VI=ab_param[1], L=L, R=Ri)
    return V


def kinetic_offdiag(n):
    # Kinetic energy matrix off-diagonal elements
    T = 2 * np.power(-1., n) / n**2
    return T


def Tmat_1d(n, dx, mtV0):
    # Kinetic energy matrix for 1-dim
    # Off-diagonal part
    T = np.arange(-n, n + 1).reshape(1, -1)
    T = T - T.transpose()
    # Force diagonals to be 1, to avoid warning on 0-dived-0, but this value assignment can be ignored as the diagonal entries are to be replaced in the next step
    T[np.diag_indices(2 * n + 1)] = 1
    T = kinetic_offdiag(T)
    # Diagonal part
    T[np.diag_indices(2 * n + 1)] = np.pi**2 / 3
    # get kinetic energy in unit of V0, mtV0 = m * V0
    T *= hb**2 / (2 * dx**2 * mtV0)
    return T


def Tmat(n, dx, mtV0):
    # Kinetic energy tensor, index order [x y z x' y' z']
    # NOTE: here n, dx are 3-element np.array s.t. n = [nx, ny, nz], dx = [dx, dy, dz]
    T = []
    for i in range(dim):
        if n[i] != 0:
            T.append(Tmat_1d(n[i], dx[i], mtV0))  # eg. T_xx'
        else:  # If the systems is set to have only 1 grid point in this direction, ie. no such dimension
            T.append(np.zeros((1, 1)))
    return T


def mem_est(n):
    n = np.array(n)
    Ndim = np.product(2 * n + 1)
    mem = 2**(2 * np.log2(Ndim) + 3 - 30)
    print("mem_set: Matrix size= {}".format(Ndim))
    print(
        "mem_set: Estimated full matrix memory usage: {:.2f} GiB".format(mem))


def H_op(T, V, n, psi0):
    N = 2 * n + 1
    psi0 = psi0.reshape(*N)
    psi = V * psi0  # delta_xx' delta_yy' delta_zz' V(x,y,z)
    psi += np.einsum('ij,jkl->ikl', T[0], psi0)  # T_xx' delta_yy' delta_zz'
    psi += np.einsum('jl,ilk->ijk', T[1], psi0)  # delta_xx' T_yy' delta_zz'
    psi += np.einsum('ij,klj->kli', T[2], psi0)  # delta_xx' delta_yy' T_zz'
    return psi.reshape(-1)


def H_mat(n,
          dx,
          avg=1,
          potential=Vfun,
          model='Gaussian',
          absorber=False,
          ab_param=(LI, VI0)):
    # Construct Hamiltonian matrix

    if model == 'Gaussian':
        mtV0 = m * V0
    elif model == 'sho':
        mtV0 = m_sho
    else:
        exit(2)

    np.set_printoptions(precision=2, suppress=True)
    print("H_mat: n={} dx={}w {} starts.".format(n, dx / w, model))
    T = Tmat(n, dx, mtV0)
    V = Vmat(n, dx, potential, float(avg), absorber, ab_param)
    return T, V


def H_solver(n,
             dx,
             k=k,
             avg=1,
             potential=Vfun,
             model='Gaussian',
             absorber=False,
             ab_param=(LI, VI0)):
    # Solve Hamiltonian matrix
    # avg factor is used to control the time average potential strength

    T, V = H_mat(n, dx, avg, potential, model, absorber, ab_param)
    mem_est(n)

    def applyH(psi):
        return H_op(T, V, n, psi)

    t0 = time()
    N = np.product(2 * n + 1)
    H = LinearOperator((N, N), matvec=applyH)
    if absorber:
        E, W = sla.eigs(H, k, which='SA')
    else:
        E, W = sla.eigsh(H, k, which='SA')
    t1 = time()
    if avg > 0:
        print('H_solver: {} Hamiltonian solved. Time spent: {:.2f}s.'.format(
            model, t1 - t0))
    elif avg == 0:
        print(
            'H_solver: free particle Hamiltonian solved. Time spent: {:.2f}s.'.
            format(t1 - t0))
    return E, W


def N_convergence(N, R, k=k, avg=1, dim=3, wavefunc=False, level=1):
    # Convergence of energy vs N, to reproduce Fig. 5a in PRA paper
    E = np.array([]).reshape(0, k)
    dim_factor = np.zeros(3, dtype=int)
    dim_factor[:dim] = 1
    R = R * dim_factor
    if wavefunc:
        x = np.linspace(-1.1 * R[0], 1.1 * R[0], int(1000))[:, None]
        p = []

    def psi(n, dx, W, x):
        V = W.reshape(
            *(np.append(2 * n + 1, -1)
              ))[:, n[1],
                 n[2], :]  # Get y=z=0 cross section of the wavefunction
        xn = np.arange(-n[0], n[0] + 1)[None] * dx[0]
        psi = 1 / np.sqrt(np.product(dx[:dim])) * np.sinc((x - xn) / dx[0]) @ V
        return psi

    for i in N:
        n = i * np.array([1, 1, 1])  # + np.array([0, 0, 3])
        dx = R / n
        n = n * dim_factor
        V, W = H_solver(n, dx, k, avg)
        E = np.append(E, V[:k].reshape(1, -1), axis=0)
        if wavefunc:
            p.append(psi(n, dx, W, x)[:, :level])
    dE = np.diff(E, axis=0)

    if wavefunc:
        return np.array(N), dE, E, x / R[0], p
    else:
        return np.array(N), dE, E


def R_convergence(N, dx):
    # Convergence of energy vs R, to reproduce Fig. 5b in PRA paper
    E = np.array([]).reshape(0, k)

    for i in N:
        Nlist = i * np.array([1, 1, 2]) + np.array([0, 0, 1])
        V, W = H_solver(Nlist, dx)
        E = np.append(E, V[:k].reshape(1, -1), axis=0)
    dE = np.diff(E, axis=0)
    R = np.array(N) * dx[0] / w
    return R, dE, E
